### ISSUE TITLE
problem with RTL languages

### ISSUE THREAD

**Issue by liordon** [HabitRPG/habitica#3465]
hey guys!
I've been working on translating HabitRPG to hebrew and I was glad to see it released today. But there is a tiny problem. Since the text on the site is usually aligned to the left and hebrew is a Right-To-Left language, it causes some disturbances in the text flow when both english and hebrew appear together. for example:

אני כותב עכשיו משהו על HabitRPG והכל יוצא לי הפוך.

In the sentence above, the word HabitRPG appears in the middle, the begginning of the sentence appears to the left of it, and the end is to the right. this means If someone is to read that sentence, he must start from the left of the word HabitRPG, read all the way left, then read the word HabitRPG itself, go all the way to the right, and read the rest from there.

TL;DR: left alignment screw up some languages, I hope you could help me with that, or maybe even I could help me :)
@paglias 

thanks in advance!
Nailo
##

**Comment by paglias** [comment]
The rest of the text is ok or would it need some particular alignment: there are problems only with Latin alphabet or is more widespread?

**Comment by paglias** [comment]
Also, can you point me on transifex to the string you used as example?

**Comment by liordon** [comment]
I encountered it only with latin alphabet since I don't use any other, but this probably happens with any LTR language. Usually, aligning the entire text to the right solves this type of problem.

This specific example is made up, but there are a lot of similar cases in the front.JSON part of the translation (this error does not appear in transifex itself since the site aligns hebrew to the right on it's own)

![capture](https://cloud.githubusercontent.com/assets/4543336/2978023/3e46f070-dbb3-11e3-84bb-bad96ee2e4c3.PNG)

^ This one is an example from the site. it's that little box in the tavern that tells you to report bugs here on github.
if there is a simple way to define the alignment of text to be Right To Left for the problematic languages, it should fix this problem for them.

**Comment by liordon** [comment]
a friend who is a bit more knowledgable than me fixed it staticly by applying the dir="RTL" command to the site's HTML. as follows:
<div class="panel panel-default" dir="RTL">
![image](https://cloud.githubusercontent.com/assets/4543336/3014183/70243086-df54-11e3-9d12-078dc22d7ddd.png)

as you can see, this quick fix isn't perfect either, but I hope it helps!

**Comment by Sardtok** [comment]
If HTML can be embedded in the translations, then you could try using dir="rtl" in combination with bdi tags (bi-directional) on the embedded English terms. You could test this with the web developer interface you have up first, to ensure that it works. If you have a local server, that would also be good to test it on.

Things like the chat can be hard to fix, because as you show, there's some messages in English and some in other languages.

There might also be a need for changes in the CSS and HTML to properly support RTL languages (this is one of the major issues with creating multi-lingual websites and software that are to support RTL languages).

**Comment by paglias** [comment]
I will try to look at it but if it's too much work I can't give any
assurance.

As far as you know, the dir tag can be applied to the body element or has
to be applies to every single element in the page?
Il 19/mag/2014 22:19 "Sigmund Hansen" notifications@github.com ha scritto:

**Comment by Sardtok** [comment]
As far as I know the dir attribute should be possible to apply to the top element (it's a global attribute):

```
<body dir="rtl">
    Right to left text here
    <bdi>some exception that is left to right</bdi>
    and more right to left text.
</body>
```

Or the other way around, wrap RTL text in bdi tags. Or you can use dir="rtl" and dir="ltr" to change between the two. There's also a CSS rule available for doing the same as the bdi tag: unicode-bidi: isolate;

See:
https://developer.mozilla.org/en-US/docs/Web/HTML/Element/bdi
https://developer.mozilla.org/en-US/docs/Web/HTML/Element/bdo

**Comment by liordon** [comment]
@paglias I know your busy in the next couple of days and can't comment, I just post this before I forget.

I've finally gathered enough courage to try @Sardtok 's advice myself, and put the HTML injection in the translation segments. I tried it for size with the DDoD, and this is the outcome:

![capture](https://cloud.githubusercontent.com/assets/4543336/3612221/c3091ea2-0da7-11e4-8ea8-31145c820cc7.PNG)

As you can see, the HTML tag works wonderfully in the description, but shows up as plaintext in the header. Is there any rule of thumb that I could use here, to know what's alignable and what is'nt?
For now I'll restrict myself to long pieces of text when experimenting.

BTW, I don't expect to turn the entire site RTL, just to fix the text boxes and such ^^"

EDIT: I looked into the code and transifex a little bit, and I think this is what I found out: most translated strings are put into the site's code between qoutation marks (") and inside predefined paragraphs (<p>...</p>), and appear as text. there are only 2 strings or so (the DDoD and Vice pt.1 texts) that come with the paragraph tags built into the string instead of the site. IDK if it helps to know, I'll try so check if transifex has a way to deal with these things or if I can insert  my own paragraph tags anyway.

**Comment by Sardtok** [comment]
I found that CSS3 has language selectors built in. This means, you could use pseudo-classes to handle whether something is going to be RTL or LTR. There would be some requirements on the code used to generate HTML.

If the body's language code is set to "ar", it will be inherited by all other tags. Some tags might need to be set to lang="en" manually in the code, though: HP and mana bars for instance. Then everything that's not specifically LTR, will be RTL for languages that fit. A decent guide is:

http://www.w3.org/International/questions/qa-css-lang

This means the body could have the lang attribute set to the language selected by the user automatically (either by the javascript or on the server side, but I'm guessing HabitRPG goes for as much work as possible being done on the client). The HTML for some of the special content, like HP bars, could have a fixed language code. Since the system is already aware of language selection, adding one line to change the body tags language in the JS source, and adding a couple of lang attributes to a select few elements in the HTML should also not take too much work. HTML injection wouldn't be necessary at all. Then the pseudo-class :lang="ar" could have its direction set to rtl: { direction: rtl; } (you could also set appropriate fonts). Other RTL languages like Hebrew and Persian/Farsi could be added as well (although font selection might have to be isolated, while the directionality is common).

For advanced text flows, there's also:

http://dev.w3.org/csswg/css-writing-modes/

Which among other things covers bi-directionality, that is when LTR text is embedded in RTL text. I came across this when I was investigating vertical writing for Japanese text.

**Comment by liordon** [comment]
Wow @Sardtok ! thanks a lot! This is immensly helpful (and really interesting)! :smiley:

**Comment by ghost** [comment]
FYI @paglias @veeeeeee 

This issue is once again brought up in the i18n guild by @_Racobot_ and @_eburcat_. Let me copy parts of the discussion:

> **_Racobot**_ I ... noticed that in my language (Hebrew), the name Habitica was always written in Hebrew. The Translation Card states that Habitica's name should be written as is. So, isn't that bad? ... I just noticed that in different resources it's actually written in English...
> 
> **_Sphinx**_ ... Translators are asked to keep Habitica untranslated. Thank you for mentioning inconsistencies! Please change translations where you can. If a particular string is reviewed and locked from further changes, please leave a comment marking it as issue.
> 
> **_eburcat**_ Well, since @_Racobot_ raised an issue with me - I'd like to hear what you think. We have a problem when we mix English words in Hebrew text. Sometimes it works, but sometimes it messes the whole sentence up. Especially when the sentences have punctuation, parentheses, etc. I'm concerned that if we stick to the policy of not translating the name Habitica - we'll completely mess up the Hebrew translation. So I'm not sure how to tackle this. Can we get an exemption for RTL languages?? I understand it's more of a legal issue - as I read somewhere the source for this policy was some kind of a trademark issue?

I believe the issue needs a fix... one way or another =)

**Comment by Sardtok** [comment]
Can the translation texts contain markup? If so, I would suggest going through translation texts and marking specific words, like Habitica, with markup setting the directional property: `Blah blah blah <bdi lang="en">Habitica</bdi>, blah blah blah.`

EDIT:
This should clearly mark the Habitica as a string that should not be translated. And should allow mixing of RTL and LTR, although it's not necessarily a flawless system (browser implementations may not be perfect).

**Comment by ghost** [comment]
Hey, I'm one of the new members in the Hebrew translation team.

I just wanted to hop in and say that adding `dir="rtl"` as an HTML body tag seems to fix any of the alignment issues. The only drawback is that doing so slightly messes with the site's design, and that it also re-aligns chats, which isn't what we want. I don't know how to fix the design problems this causes, but applying a `dir="ltr"`tag individually to the chat element seems to fix that one.

I have to add, I don't really have any knowledge in programming. I simply used Google Chrome's "Inspect element" feature for testing this.

Also, here are some screenshots of the changes caused by adding the dir tag - http://imgur.com/a/q17k6

The most noticeable is the player bar. Almost everything changes it's place. To be honest, I think this is actually a good outcome, since it's (obviously) a RTL language. but it's far from perfect.
The most major problem I believe is with images and other elements that have fixed positions. You can see in the Pets page that the foods box hides some of the pets, and that Matt's picture and chatbox stayed in the same place (when it should ideally move to the right side).

As for the markup Sardtok mentioned, I honestly have no idea what that is. However, I did try to include `<bdi lang="en">Habitica</bdi>` in one of the translations, but it didn't seem to update on the actual site. Perhaps a reviewer has to approve the change before it goes live?

**Comment by Sardtok** [comment]
bdi is a tag for isolating text with a possibly different directionality in bidirectional text. https://developer.mozilla.org/en/docs/Web/HTML/Element/bdi

To work properly, it might require directionality set in CSS, though. There's also the bdo tag which allows forcing directionality, but I think the dir="rtl" already forces the directionality if there are no overrides anywhere.

**Comment by ghost** [comment]
I was asked by Alys to describe an issue caused by this in the chat area.

So basically, when posting a message in an RTL language, it jumbles up the timestamp. Instead of just showing "5 days ago", for example, it shows "5" on one side of the message and "days ago" on the other.

And just as anywhere else in the site - if the message includes both English and RTL words, it screws it all up as originally described by @liordon 

Here is an example showing what I wrote in Hebrew and how it appeared in the chat (the bottom message is written only in Hebrew, while the top one also includes English):
![1](http://i.imgur.com/ZS4MjiL.png)
![2](http://i.imgur.com/fxZZa0Q.png)
So in the top message you can see that the words are completely mixed up, and so is the timestamp. And in the bottom message, the timestamp is also jumbled up, and when hovering over it using the mouse, the popup appears in the middle of the message, instead of right above the "days ago".

**Comment by Alys** [comment]
I've marked this issue as "interface" because we might be able to improve that situation if we decide to modify the chat code in any future redesign of the website.

Here's the problem described by Racobot in context from https://habitica.com/#/options/groups/guilds/17d57080-1bb2-407a-8276-63282e44c3ce :

![image](https://cloud.githubusercontent.com/assets/1495809/16373972/d363a63e-3c97-11e6-95a0-c8150d1bbe44.png)

**Comment by paglias** [comment]
Putting this issue on hold until the redesign but if you have any suggestion, feel free to write about it here

**Comment by lemoness** [comment]
We still have some issues with this, but are they different enough from the original problems that we need to open a new ticket, or can we keep this one?

**Comment by citrusella** [comment]
Out of curiosity: *Is* this still occurring? What are the "some issues" mentioned just after the redesign? When I try to sort of force what's happening in the OP to happen in a PM to myself (by adding a random English word into a string of Hebrew), it seems to work the way I expect it from seeing things on other websites? (AFAICT, as I'm not fluent in Hebrew.)

**Comment by SabreCat** [comment]
Good question. I'm going to close this for now as we haven't had any commentary on it in a very long time, and what commentary does exist is pretty vague. If we get new reports of specific trouble with RTL, we can write it up more precisely!
